rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,283 @@
+  %1 = "stablehlo.add"(%arg0, %arg2) : (tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>, tensor<?xf32, #stablehlo.type_extensions<bounds = [4]>>) -> tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>
+  %2 = "stablehlo.add"(%arg0, %arg3) : (tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>, tensor<2xf32>) -> tensor<2xf32>
+
+  // This is illegal as operands have incompatible types. %arg0 can either be
+  // tensor<0xf32>, tensor<1xf32>, tensor<2xf32> or tensor<3xf32> at runtime,
+  // none of these are compatible with tensor<4xf32>
+  %3 = "stablehlo.add"(%arg0, %arg4) : (tensor<?xf32, #stablehlo.type_extensions<bounds = [3]>>, tensor<4xf32>) -> tensor<*xf32>
+  func.return
+}
+```
+
+Currently, the StableHLO dialect uses the MLIR ranked tensor type to represent
+bounds. In the future we plan to introduce StableHLO type which supports bounds,
+along with a custom pretty printing format.  There will be a separate RFC on
+this. Also, the proposal will follow StableHLO backward compatibility policies
+so it is safe to use `TypeExtensionsAttr` currently.
+
+### (P2) StableHLO op semantics with bounded operands or results
+
+All ops that support dynamic operands or results can have bounds specified for
+them. However, the result types need to be compatible with the inferred result
+types. This allows result types to be more generic or specific as long as it is
+compatible with the inferred type.
+
+Separately, the StableHLO specification will be updated to cover bounded types
+for all the relevant ops.
+
+### (P3) get\_dimension\_size / set\_dimension\_size ops
+
+The `get_dimension_size` op takes a tensor and a dimension index and returns the
+runtime size as `tensor<i32>`.
+
+The following example returns the size of the result after concatenating input
+that has up to `16` elements with self and returns the runtime size of the
+concatenation result.
+
+```mlir
+func.func @self_concat_size(%data: tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>) -> tensor<i32> {
+  %concat = "stablehlo.concatenate"(%data, %data) {dimension = 0 : i64}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>,
+       tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>)
+      -> tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>
+
+  %result = stablehlo.get_dimension_size %concat, dim = 0
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>) -> tensor<i32>
+
+  func.return %result : tensor<i32>
+}
+```
+
+The `set_dimension_size` op takes a static or bounded tensor, runtime size and a
+dimension index and returns a tensor whose size of the particular dimension is
+set to the specified size. This size needs to be less than or equal to the
+static size or bound for the dimension. This operation can be thought as either
```

Comment:
Shouldn't it is like the runtime size needs to be <= just the bounds (not the static size). This is because we do have a semantics, for a slice, when the dimension size is bigger than runtime size.

I am confused while reading this with the following sentence.

Reply:
Here is the relevant check in HLO. It doesn't check if the dimension is static or bounded.
https://github.com/openxla/xla/blob/b809f5b6eebcf30c7d54fcbd23abbee97bdba1be/xla/literal.cc#L709

I don't understand the confusion. could you expand?

Reply:
From our offline chat + the above link the confusion is cleared.